horovod/mxnet/mpi_ops.cc

```diff
@@ -49,206 +51,154 @@ inline void InvokeCompleteCallback(CallbackOnComplete on_complete, const Status&
-      exec_fn = [input, output, op_name]
-                (RunContext rctx, CallbackOnComplete on_complete) mutable {
-        DoAllreduce(input, output, op_name, on_complete);
-      };
+      hvd_output = std::make_shared<MXTensor<NDArray>>(output);
+      enqueue_result = EnqueueTensorAllreduce(
+          hvd_context, hvd_tensor, hvd_output, nullptr, name, device,
+          [on_complete](const Status& status) {
+            InvokeCompleteCallback(on_complete, status);
+      });
       break;
     case OperationType::ALLGATHER:
-      op_type_name = "horovod_allgather";
-      op_name = GetOpName(op_type_name, name);
-      exec_fn = [input, output, op_name]
-                (RunContext rctx, CallbackOnComplete on_complete) mutable {
-        DoAllgather(input, output, op_name, on_complete);
-      };
+      enqueue_result = EnqueueTensorAllgather(
+          hvd_context, hvd_tensor, nullptr, name, device,
+          [on_complete](const Status& status) {
+            InvokeCompleteCallback(on_complete, status);
+      });
       break;
     case OperationType::BROADCAST:
-      op_type_name = "horovod_broadcast";
-      op_name = GetOpName(op_type_name, name);
-      exec_fn = [input, output, op_name, root_rank]
-                (RunContext rctx, CallbackOnComplete on_complete) mutable {
-        DoBroadcast(input, output, op_name, root_rank, on_complete);
-      };
+      if (horovod_rank() == ops_param->root_rank) {
+        if (tensor != output) {
+          TensorUtil::Copy(output, tensor);
+        }
+      } else {
+        hvd_output = std::make_shared<MXTensor<NDArray>>(output);
+      }
+
+      enqueue_result = EnqueueTensorBroadcast(
+          hvd_context, hvd_tensor, hvd_output, ops_param->root_rank,
+          nullptr, name, device,
+          [on_complete](const Status& status) {
+            InvokeCompleteCallback(on_complete, status);
+      });
       break;
     default:
-      LOG(FATAL) << "Unsupported Horovod operation type";
+      throw std::logic_error("Unsupported Horovod operation type.");
   }
 
+  ThrowIfError(enqueue_result);
+}
+
+inline void PushHorovodOperation(OperationType op_type, NDArray* input,
+                                 NDArray* output, const char* name,
+                                 int priority, int root_rank = -1) {
+  auto op_type_name = GetOpTypeName(op_type);
+  auto op_name = GetOpName(op_type_name, name);
+  auto ops_param = CreateMpiOpsParam(input, output, op_type, op_name, root_rank, false);
+
   // Not in-place
-  if (input->var() != output->var()) {
-    Engine::Get()->PushAsync(exec_fn, MX_EXEC_CTX,
-                             {input->var()}, {output->var()},
-                             FnProperty::kCPUPrioritized, priority,
-                             op_type_name.c_str());
+  auto input_var = input->var();
+  auto output_var = output->var();
+  if (input_var != output_var) {
+    MXEnginePushAsync(DoHorovodOperation, ops_param, DeleteMpiOpsParam,
```

Comment:
What happens if an user installed mxnet-1.4 instead of mxnet nightly? Does hovorod plan to drop mxnet 1.4 support? 
(I'm fine with mxnet nightly for my project, but I cannot speak for others)

Reply:
Users can still use Horovod 0.16.0 and 0.16.1 releases with MXNet 1.4.0 release. Given that GCC incompatibility issue we had in earlier releases, I would encourage users to upgrade to latest Horovod and MXNet releases. Maintaining two internal implementations here is cumbersome.

Reply:
ok